server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java

```diff
@@ -0,0 +1,272 @@
+        assertThat(countRequestsForIndex(state, index), is(equalTo(0)));
+    }
+
+    public void testMixOfStartedPromotableAndSearchReplicas() throws Exception {
+        final String index = "test";
+        final int numShards = 1 + randomInt(3);
+        final int numPromotableReplicas = randomInt(2);
+        final int numSearchReplicas = randomInt(2);
+
+        ClusterState state = stateWithAssignedPrimariesAndReplicas(
+            new String[] { index },
+            numShards,
+            getReplicaRoles(numPromotableReplicas, numSearchReplicas)
+        );
+        setState(clusterService, state);
+        logger.debug("--> using initial state:\n{}", clusterService.state());
+        assertThat(countRequestsForIndex(state, index), is(equalTo(numShards * numSearchReplicas)));
+    }
+
+    public void testSearchReplicasWithRandomStates() throws Exception {
+        final String index = "test";
+        final int numPromotableReplicas = randomInt(2);
+        final int numSearchReplicas = randomInt(6);
+
+        List<Tuple<ShardRoutingState, ShardRouting.Role>> replicas = getReplicaRolesWithRandomStates(
+            numPromotableReplicas,
+            numSearchReplicas
+        );
+        int numReachableUnpromotables = replicas.stream().mapToInt(t -> {
+            if (t.v2() == ShardRouting.Role.SEARCH_ONLY && t.v1() != ShardRoutingState.UNASSIGNED) {
+                if (t.v1() == ShardRoutingState.RELOCATING) {
+                    return 2; // accounts for both the RELOCATING and the INITIALIZING copies
+                }
+                return 1;
+            }
+            return 0;
+        }).sum();
+
+        final ClusterState state = state(index, true, ShardRoutingState.STARTED, replicas);
+
+        setState(clusterService, state);
+        logger.debug("--> using initial state:\n{}", clusterService.state());
+        assertThat(countRequestsForIndex(state, index), is(equalTo(numReachableUnpromotables)));
+    }
+
+    public void testInvalidPrimaryShard() throws Exception {
+        final String index = "test";
+        ClusterState state = stateWithAssignedPrimariesAndReplicas(
+            new String[] { index },
+            randomIntBetween(1, 3),
+            getReplicaRoles(randomInt(2), randomInt(2))
+        );
+        setState(clusterService, state);
+        logger.debug("--> using initial state:\n{}", clusterService.state());
+
+        ShardId shardId = state.routingTable().activePrimaryShardsGrouped(new String[] { index }, true).get(0).shardId();
+        ShardId wrongShardId = new ShardId(shardId.getIndex(), shardId.getId() + randomIntBetween(10, 100));
+        PlainActionFuture<ActionResponse.Empty> response = PlainActionFuture.newFuture();
+        logger.debug("--> executing for primary shard id: {}", wrongShardId);
+        ActionTestUtils.execute(broadcastUnpromotableAction, null, new TestBroadcastUnpromotableRequest(wrongShardId), response);
```

Comment:
nit: maybe use the static `PlainActionFuture#get` for things like this (ideally one with a timeout, there's no need to wait forever here)

Reply:
It's done in the line below with `response.get()`. Feel free to tell me if this is not what you meant.

Reply:
Feel free to unresolve the conversation.

Reply:
I mean the _static_ `get` method:

```
        logger.debug("--> executing for primary shard id: {}", wrongShardId);
        assertThat(
            expectThrows(
                ShardNotFoundException.class,
                () -> PlainActionFuture.<ActionResponse.Empty, Exception>get(
                    f -> ActionTestUtils.execute(broadcastUnpromotableAction, null, new TestBroadcastUnpromotableRequest(wrongShardId), f),
                    10,
                    TimeUnit.SECONDS
                )
            ).toString(),
            allOf(containsString("no such shard"), containsString(wrongShardId.toString()))
        );
```